### PR TITLE
Rudimentary docker network support

### PR DESCRIPTION
**Pull request by dcbw**
Start adding support for docker's network API (which is just a passthrough to libnetwork's API).

### CODE REVIEW

**Comment by peter-edge** [comment]
Hey, just for reference, can you link the github.com/docker/docker corresponding code/docs? If you don't have time, I can check it out soon.

File: network.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+// Service represents a service.
+type Service struct {
+	ID string `json:"id"`
+}
+
+// ListNetworks returns all networks.
+func (c *Client) ListNetworks() ([]Network, error) {
+	path := "/networks"
```
Comment (peter-edge) [inline network.go]: nit: might as well just inline this :)

File: network.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	err = json.Unmarshal(body, &networks)
+	if err != nil {
+		return nil, err
+	}
+	return networks, nil
+}
+
+// NetworkInfo returns information about a network by its ID.
+func (c *Client) NetworkInfo(id string) (*Network, error) {
+	path := "/networks/" + id
```
Comment (peter-edge) [inline network.go]: same comment

File: network.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+// ListNetworks returns all networks.
+func (c *Client) ListNetworks() ([]Network, error) {
+	path := "/networks"
+	body, _, err := c.do("GET", path, doOptions{})
+	if err != nil {
+		return nil, err
+	}
+	var networks []Network
+	err = json.Unmarshal(body, &networks)
```
Comment (peter-edge) [inline network.go]: scope this (and below):

```
if err := json.Unmarshal(body, &networks); err != nil {
  return nil, err
}
```

File: network.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return &network, nil
+}
+
+// NoSuchNetwork is the error returned when a given network does not exist.
+type NoSuchNetwork struct {
+	ID  string
+	Err error
```
Comment (peter-edge) [inline network.go]: do you ever use Err? it doesn't look like it here.

File: network.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+func (err *NoSuchNetwork) Error() string {
+	if err.Err != nil {
+		return err.Err.Error()
+	}
+	return "No such network: " + err.ID
```
Comment (peter-edge) [inline network.go]: nit: i usually like `fmt.Sprintf("foo: %s", err.ID)` but no biggie

**Comment by peter-edge** [comment]
Also overall: needs some testing :)

File: container.go [hunk 6]
```diff
@@ -135,6 +135,8 @@ type NetworkSettings struct {
 	Bridge      string                 `json:"Bridge,omitempty" yaml:"Bridge,omitempty"`
 	PortMapping map[string]PortMapping `json:"PortMapping,omitempty" yaml:"PortMapping,omitempty"`
 	Ports       map[Port][]PortBinding `json:"Ports,omitempty" yaml:"Ports,omitempty"`
+	NetworkID   string                 `json:"NetworkID,omitempty" yaml:"NetworkID,omitempty"`
```
Comment (peter-edge) [inline container.go]: just making sure - is it `json:"NetworkID"` or `json:"NetworkId"` for docker?

Comment (dcbw) [inline container.go]: It's "NetworkID":

        "MacAddress": "02:42:ac:11:2a:02",
        "NetworkID": "b6408b405133c5475bfde2ad383b80732326e1fb9bc713f6586552836ba2fdd5",
        "PortMapping": null,
        "Ports": {},
        "SandboxKey": "/var/run/docker/netns/23659bc4a779",
        "SecondaryIPAddresses": null,
        "SecondaryIPv6Addresses": null

File: network.go (hunk 4 above)
Comment (dcbw) [inline network.go]: I mainly copied NoSuchContainer, which also has Err and doesn't appear to use it.  I'm happy to take it out of NoSuchNetwork, but should I be consistent with NoSuchContainer, or should I remove the unused bit?

Comment (peter-edge) [inline network.go]: Remove both, thanks :)

File: container.go (hunk 6 above)
Comment (peter-edge) [inline container.go]: SGTM

File: network.go (hunk 1 above)
Comment (dcbw) [inline network.go]: Sure.  I removed Service too, since it's currently unused.

**Comment by dcbw** [comment]
Repushed with some fixups.

**Comment by peter-edge** [comment]
Thanks! Still needs to address the string inlining, linking to docker/docker code, and tests :)

**Comment by dcbw** [comment]
FWIW the JSON looks like this:

**Comment by dcbw** [comment]
FWIW the JSON for Network looks like this:

&{Name:purple ID:2585945e1bb6f36514dde1c563dd8d078bc058d9a4953a0fd8780b3b5237e983 Type:cni Endpoints:[]}

**Comment by dcbw** [comment]
Repushed with inlined string and added links to docker code.  No documentation yet on the API that I can find (not surprising, it's really new).  Still have to do some tests; but does the code look OK so far now?

**Comment by fsouza** [comment]
@peter-edge thanks for reviewing this!

@dcbw except for the lack of tests, I think the code looks good for now. Regarding the Docker documentation issue, are these APIs available in Docker 1.8? If so, they're probably going to add some docs soon.

**Comment by dcbw** [comment]
@fsouza these APIs are available in docker 1.8 if you enable the EXPERIMENTAL flag when building docker.  It seems very likely they'll get turned on by default in the near future though.

**Comment by peter-edge** [comment]
@fsouza i think it's good to merge but could use at least some base for testing so that when this code is inevitably expanded, other OSS contributors know where to add tests. plus some documentation about the need to do EXPERIMENTAL/1.8 etc. @dcbw are you willing to do this? :)

**Comment by dcbw** [comment]
@peter-edge yes, I'm willing to work on the tests in the next few days.

**Comment by fsouza** [comment]
Cool, we can wait for the tests then!

@dcbw thank you!

**Comment by dcbw** [comment]
Ok, added some testcases, do they look complete enough?  Also added a note to the README about the network API and DOCKER_EXPERIMENTAL, and added some more fields to the NetworkSettings struct that are present in docker 1.8.

**Comment by peter-edge** [comment]
Looks great! @fsouza

**Comment by fsouza** [comment]
I'm missing the tests for the CreateNetwork method. @dcbw any special reason for not having tests for this method?

Thank you! And sorry for the late response.

**Comment by dcbw** [comment]
@fsouza no, just an oversight, sorry!  Added the client-side CreateNetwork tests and repushed.

**Comment by fsouza** [comment]
@dcbw thank you!

**Pull request merged**